sklearn/utils/validation.py

```diff
@@ -80,7 +80,12 @@ def as_float_array(X, copy=True, force_all_finite=True):
     elif X.dtype in [np.float32, np.float64]:  # is numpy array
         return X.copy('F' if X.flags['F_CONTIGUOUS'] else 'C') if copy else X
     else:
-        return X.astype(np.float32 if X.dtype == np.int32 else np.float64)
+        if X.dtype.kind in 'uib' and \
+           X.dtype.itemsize <= np.int32(0).dtype.itemsize:
```

Comment:
Is `np.int32(0).dtype.itemsize` always `4` whatever the platform? I would guess so.